src/github.com/matrix-org/dendrite/encryptoapi/routing/keys.go

```diff
@@ -0,0 +1,537 @@
+) (al []string, err error) {
+	al, err = encryptDB.SelectAl(ctx, uid, device)
+	return
+}
+
+func pickOne(
+	ctx context.Context,
+	encryptDB storage.Database,
+	uid, device, al string,
+) (key types.KeyHolder, err error) {
+	key, err = encryptDB.SelectOneTimeKeySingle(ctx, uid, device, al)
+	return
+}
+
+func upnotify(userID string) {
+	m := sarama.ProducerMessage{
+		Topic: "keyUpdate",
+		Key:   sarama.StringEncoder("key"),
+		Value: sarama.StringEncoder(userID),
+	}
+	keyProducer.ch.Input() <- &m
+}
+
+// InitNotifier initialize kafka notifier
+func InitNotifier(base *basecomponent.BaseDendrite) {
+	keyProducer.base = base
+	pro, _ := sarama.NewAsyncProducer(base.Cfg.Kafka.Addresses, nil)
+	keyProducer.ch = pro
+}
+
+func presetDeviceKeysQueryMap(
+	deviceKeysQueryMap map[string]types.DeviceKeysQuery,
+	uid string,
+	key types.KeyHolder,
+) map[string]types.DeviceKeysQuery {
+	// preset for complicated nested map struct
+	if _, ok := deviceKeysQueryMap[key.DeviceID]; !ok {
+		// make consistency
+		deviceKeysQueryMap[key.DeviceID] = types.DeviceKeysQuery{}
+	}
+	if deviceKeysQueryMap[key.DeviceID].Signature == nil {
+		mid := make(map[string]map[string]string)
+		midmap := deviceKeysQueryMap[key.DeviceID]
+		midmap.Signature = mid
+		deviceKeysQueryMap[key.DeviceID] = midmap
+	}
+	if deviceKeysQueryMap[key.DeviceID].Keys == nil {
+		mid := make(map[string]string)
+		midmap := deviceKeysQueryMap[key.DeviceID]
+		midmap.Keys = mid
+		deviceKeysQueryMap[key.DeviceID] = midmap
+	}
+	if _, ok := deviceKeysQueryMap[key.DeviceID].Signature[uid]; !ok {
+		// make consistency
+		deviceKeysQueryMap[key.DeviceID].Signature[uid] = make(map[string]string)
+	}
+	return deviceKeysQueryMap
+}
+
+func bothKeyProcess(
+	ctx context.Context,
+	body *types.UploadEncryptSpecific,
+	userID, deviceID string,
+	database *storage.Database,
+	deviceKeys types.DeviceKeys,
+) (err error) {
+	// insert one time keys firstly
+	onetimeKeys := body.OneTimeKey
+	for alKeyID, val := range onetimeKeys.KeyString {
+		al := (strings.Split(alKeyID, ":"))[0]
+		keyID := (strings.Split(alKeyID, ":"))[1]
+		keyInfo := val
+		keyStringTyp := ONETIMEKEYSTR
+		sig := ""
+		err = database.InsertKey(ctx, deviceID, userID, keyID, keyStringTyp, keyInfo, al, sig)
+		if err != nil {
+			return
+		}
+	}
+	for alKeyID, val := range onetimeKeys.KeyObject {
+		al := (strings.Split(alKeyID, ":"))[0]
+		keyID := (strings.Split(alKeyID, ":"))[1]
+		keyInfo := val.Key
+		keyObjectTyp := ONETIMEKEYSTR
+		sig := val.Signature[userID][fmt.Sprintf("%s:%s", "ed25519", deviceID)]
+		err = database.InsertKey(ctx, deviceID, userID, keyID, keyObjectTyp, keyInfo, al, sig)
+		if err != nil {
+			return
+		}
+	}
+	// insert device keys
+	keys := deviceKeys.Keys
+	sigs := deviceKeys.Signature
+	for alDevice, key := range keys {
+		al := (strings.Split(alDevice, ":"))[0]
+		keyTyp := DEVICEKEYSTR
+		keyInfo := key
+		keyID := ""
+		sig := sigs[userID][fmt.Sprintf("%s:%s", "ed25519", deviceID)]
```

Comment:
I believe `"ed25519:" + deviceID` is simpler. Is there a reason for `fmt.Sprintf` ?

Reply:
fmt is a docked-in pkg in go , and it takes the same overhead as "+" but a better format restrict , in consideration of the protocol would be changed one day , I think it's more flexible.